feat(expo-router): allow specifying user-defined headers for all routes

hassankhan opened the pull request:
# Why

This change allows users to specify custom headers that apply to all routes via a new `headers` property in `expo-router`'s config plugin options.

# How

In development, `createRouteHandlerMiddleware()` applies the user-defined headers in `createRequestHandler()`'s `beforeResponse` hook.

In production, we add the user-defined headers to the serialized server manifest when exporting the app. When the exported app is served, we apply the headers in `createEnvironment()`'s `beforeResponse` hook.

Notably, values defined in `headers` __do not__ apply to redirects, since [`Response.redirect()`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirect_static) ([used here](https://github.com/expo/expo/blob/5ab0d1e0ec17086d33af4303d23ac6b0af5a8012/packages/expo-server/src/vendor/abstract.ts#L328)) doesn't allow for extra headers to be passed along with the redirect.

# Test Plan

- CI
- Manual testing with `cd apps/router-e2e && yarn start:headers`

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

hassankhan committed:
feat(expo-router): allow specifying user-defined headers

hassankhan committed:
feat(expo-router): allow specifying user-defined headers in production

hassankhan committed:
chore(expo-router): update changelogs

hassankhan committed:
test(expo-router): add tests for user-defined headers

hassankhan committed:
chore: add missing changelog

hassankhan committed:
chore(expo-router): improve test project invocation

hassankhan committed:
test(expo-router): tidy tests

hassankhan committed:
chore: update changelogs

hassankhan committed:
test(expo-router): improve test cases

kitten commented on packages/expo-router/plugin/options.json:
```diff
@@ -136,6 +136,23 @@
+        "headers": {
+          "description": "A list of header values that are sent with every response from the server",
+          "type": "object",
+          "additionalProperties": {
```
future: We could validate `propertyNames` here in the future, I guess, but I'm not sure if the error would be as useful. Let's leave that for a future PR maybe

kitten commented on packages/expo-server/src/vendor/environment/common.ts:
```diff
@@ -33,10 +33,30 @@ interface EnvironmentInput {
+      return manifest;
+    },
+
+    beforeResponse(responseInit: any, route: any) {
```
Do we have any reservations to moving this to `src/vendor/abstract.ts`? I don't really see this being different in any case, so moving it would be nice. My main concern here is that we'd like this to be consistent and the types of `responsInit` and `route` are consistent, but if we move it here, we're opening it up to changes across environments.

The main intention of the `before*` hooks was basically to allow the CLI to do specific things and to allow EAS Hosting to step in (since it needs additional response serialisation), so not sure we need to make this customisable and feed this back.

hassankhan replied:
No reservations as such, it just seemed like a convenient place to hook into the response lifecycle. Will address in a followup commit.

kitten commented on packages/expo-router/plugin/options.json:
```diff
@@ -136,6 +136,23 @@
           }
         },
+        "headers": {
+          "description": "A list of header values that are sent with every response from the server",
```
I think scoping this down is fine. As discussed, it's a bit unclear how to deal with "none route" responses unless we hoist this, the problem being:
- if we want this to apply in EAS Hosting to asset responses, we have to hoist this
- but, if we hoist this, we also then need to teach users how to apply this for every adapter, since adapters don't include assets (and instructions for Vercel/Netlify would differ and could get tricky)

So, overall, I'm fine with applying this to only routes that Expo Router controls, and treating assets as out-of-scope for now, (although, some hoisting could be applied there), but I'd change this description slightly to:

> A list of headers that are set on every route response from the server

To imply that it won't necessarily apply to assets.

hassankhan committed:
refactor(expo-router): apply review feedback

hassankhan committed:
fix(expo-router): throw error when `web.output` is not set to `server`

Ubax reviewed (approved):
LGTM 👍

kitten committed:
Remove `routerOptions` check

It's not `MetroBundlerDevServer`'s responsibility to check this and this
is reaching outside of what it should do. We also don't want to validate
this here since, eventually, this will make its way into the `'static'`
and `'single'` config too and we should validate this in "runners"
instead.
